### PR TITLE
raftstore: increase batch for raftlog-gc-worker

### PR DESCRIPTION
**Pull request by Little-Wallace**
Signed-off-by: Little-Wallace <bupt2013211450@gmail.com>

### What problem does this PR solve?

Issue Number: close https://github.com/tikv/tikv/issues/11404

### What is changed and how it works?

Batch more deleted keys together. As we all know ,the deleted key must be much small than the entry which has been put. So 256 is too small for every writebatch in raftlog-gc-worker.
And the gc tasks of most of regions are small. They are often smaller than 100. So I batch the keys of multiple regions and write them together.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- PR to update `pingcap/tidb-ansible`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/engine_rocks/src/raft_engine.rs [hunk 1]
```diff
@@ -97,6 +97,38 @@ impl RaftEngineReadOnly for RocksEngine {
...
+        raft_wb: &mut RocksWriteBatch,
+    ) -> Result<usize> {
+        if from >= to {
+            return Ok(0);
+        }
+        if from == 0 {
+            let start_key = keys::raft_log_key(raft_group_id, 0);
+            let prefix = keys::raft_log_prefix(raft_group_id);
+            match self.seek(&start_key)? {
+                Some((k, _)) if k.starts_with(&prefix) => from = box_try!(keys::raft_log_index(&k)),
```
Comment (5kbpers) [inline components/engine_rocks/src/raft_engine.rs:115]: Make sure `from < to` after `seek`

Comment (5kbpers) [inline components/engine_rocks/src/raft_engine.rs:115]: Make sure from < to after seek

File: components/engine_rocks/src/raft_engine.rs [hunk 2]
```diff
@@ -172,35 +204,27 @@ impl RaftEngine for RocksEngine {
...
-            let start_key = keys::raft_log_key(raft_group_id, 0);
-            let prefix = keys::raft_log_prefix(raft_group_id);
-            match self.seek(&start_key)? {
-                Some((k, _)) if k.starts_with(&prefix) => from = box_try!(keys::raft_log_index(&k)),
-                // No need to gc.
-                _ => return Ok(0),
-            }
-        }
-
+    fn batch_gc(&self, groups: Vec<(u64, u64, u64)>) -> Result<usize> {
```
Comment (5kbpers) [inline components/engine_rocks/src/raft_engine.rs:207]: Better to use a struct instead of a tuple.

**Review by 5kbpers — APPROVED** [review]

File: components/raftstore/src/store/worker/raftlog_gc.rs [hunk 3]
```diff
@@ -108,8 +105,8 @@ impl<EK: KvEngine, ER: RaftEngine, R: CasualRouter<EK>> Runner<EK, ER, R> {
         });
         RAFT_LOG_GC_KV_SYNC_DURATION_HISTOGRAM.observe(start.saturating_elapsed_secs());
         let tasks = std::mem::take(&mut self.tasks);
+        let mut groups = Vec::with_capacity(tasks.len());
```
Comment (BusyJay) [inline components/raftstore/src/store/worker/raftlog_gc.rs:108]: Better use a HashMap, so releated deletes are clustered together.

File: components/raftstore/src/store/worker/raftlog_gc.rs [hunk 4]
```diff
@@ -149,6 +135,20 @@ impl<EK: KvEngine, ER: RaftEngine, R: CasualRouter<EK>> Runner<EK, ER, R> {
                 }
             }
         }
+        let start = Instant::now();
+        match self.gc_raft_log(groups) {
```
Comment (BusyJay) [inline components/raftstore/src/store/worker/raftlog_gc.rs:139]: This reorder the `Task::Purge` and `Task::Gc`, I'm not sure if it's OK. /cc @tabokie

File: components/raftstore/src/store/worker/raftlog_gc.rs [hunk 5]
```diff
@@ -120,21 +117,10 @@ impl<EK: KvEngine, ER: RaftEngine, R: CasualRouter<EK>> Runner<EK, ER, R> {
...
-                            RAFT_LOG_GC_FAILED.inc();
-                        }
-                        Ok(n) => {
-                            debug!("gc log entries"; "region_id" => region_id, "entry_count" => n);
-                            self.report_collected(n);
-                            RAFT_LOG_GC_DELETED_KEYS_HISTOGRAM.observe(n as f64);
-                        }
-                    }
-                    RAFT_LOG_GC_WRITE_DURATION_HISTOGRAM.observe(start.saturating_elapsed_secs());
+                    groups.push((region_id, start_idx, end_idx));
```
Comment (BusyJay) [inline components/raftstore/src/store/worker/raftlog_gc.rs:120]: Should log abnormal gap and its region id.

File: components/raftstore/src/store/worker/raftlog_gc.rs (hunk 4 above)
Comment (tabokie) [reply components/raftstore/src/store/worker/raftlog_gc.rs:144]: It's fine, they are unrelated.

**Review by BusyJay — APPROVED** [review]

**Comment by sticnarf** [comment]
Does this PR need to be picked to other release branches?

**Comment by BusyJay** [comment]
It was picked to 5.1, but it should also be picked to 5.2 and 5.3.

**Pull request merged**